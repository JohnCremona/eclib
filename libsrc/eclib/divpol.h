// divpol.h: declaration of functions for division polynomials
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// allow for multiple includes
#ifndef _ECLIB_DIVPOL_
#define _ECLIB_DIVPOL_

#include "curve.h"

// These functions return the vector of coefficients (starting with
// the constant term) rather than polynomial types

// div_pol_odd(a1,a2,a3,a4,a6,n) returns the coefficients of the
// polynomial in x whose zeros are the (x-coordinates of the) non-zero
// points P on E=[a1,a2,a3,a4,a6] satisfying nP=0 (odd n)

// The poly itself is found recursively

// Despite the name, for even n this returns a correct n-division
// polynomial without the 2-torsion factor, i.e. the polynomial whoe
// roots are the x-coordinates of the points P satisfying nP=0, 2P!=0.

ZPoly div_pol_odd(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                  const ZZ& a6,int n);

ZPoly div_pol_2(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                const ZZ& a6);

ZPoly div_pol(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
              const ZZ& a6,int n);

ZPoly division_polynomial(Curvedata* EE, int p);

// Numerator and denominator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_num(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                   const ZZ& a6, int n);

ZPoly mul_by_n_den(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                   const ZZ& a6, int n);

// Polynomial whose roots are x(Q) for Q satisfying n*Q=P, where x(P)=xP/zP

ZPoly division_points_X_pol(const ZZ& a1,const ZZ& a2,const ZZ& a3,const ZZ& a4,
                            const ZZ& a6,
                            int n,
                            const ZZ& xP, const ZZ& zP);

#endif // #define _DIVPOL_
