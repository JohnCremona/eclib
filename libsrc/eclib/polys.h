// polys.h : defines interface to NTL polynomials
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2012 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////

// allow for multiple includes
#ifndef _ECLIB_POLYS_H
#define _ECLIB_POLYS_H

// NB --warning!  In NTL there is a universal modulus set by a single
// call to ZZ_p::init(q).

#include <eclib/gf.h>  // scalars
#include <eclib/bigrat.h>

#include <NTL/ZZ_pX.h>
#include <NTL/ZZ_pXFactoring.h>

#define ZPoly ZZX
#define PolyCoeff(f,i) coeff((f),(i))
#define ZPolySetX(f) SetX(f);
#define SetDegree(f,d)
#define Degree(f) deg((f))

#define FqPoly ZZ_pX
#define NewFqPoly(field,name) FqPoly name
#define FqPolySetField(f,field) // do nothing
#define FqPolyAssignGF(f,c)  f=((c))
#define FqPolyAssignZ(f,c)  f=(to_ZZ_p(c))
#define FqPolyAssign0(f)  f=to_ZZ_p(0)
#define FqPolyAssign1(f)  f=to_ZZ_p(1)
#define FqPolyEval(f,c)  eval((f),to_ZZ_p(c))
#define FqPolyAssignX(f) SetX(f)
#define GetField(f) (galois_field(ZZ_p::modulus()))

vector<bigint> rootsmod(const vector<bigint>& coeffs, bigint p);
vector<gf_element> roots(const FqPoly& f);
vector<bigrational> roots(const vector<bigint>& coeffs);
vector<bigrational> roots(const ZPoly& f);
vector<bigint> introots(const ZPoly& f);

vector<bigint> Introotscubic(const bigint& a, const bigint& b, const bigint& c);
vector<bigint> Introotsquartic(const bigint& a, const bigint& b, const bigint& c,
                            const bigint& d);



// find the number of roots of X^3 + bX^2 + cX + d = 0 (mod p)
int nrootscubic(const bigint& bb, const bigint& cc, const bigint& dd, const bigint& p);

FqPoly reduce(const ZPoly& f, const galois_field& Fq);

inline FqPoly reduce(const ZPoly& f, const bigint& q)
{return reduce(f,galois_field(q));}



#endif // #define _POLYS_
