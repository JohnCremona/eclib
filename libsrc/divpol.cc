// divpol.cc: implementations of functions for division polynomials
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
 
// NB the external interface uses a simple vector<bigint> of
// coefficients rather than a polynomial type to simplify the
// NTL interface

#include <eclib/polys.h>
#include <eclib/divpol.h>

// The 2-divison polynomial (cubic in x)

ZPoly div_pol_2(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                const bigint& a6)
{
  ZPoly ans;
  SetDegree(ans,3);
  SetCoeff(ans,3,4);
  SetCoeff(ans,2,a1*a1+4*a2);
  SetCoeff(ans,1,2*a1*a3+4*a4);
  SetCoeff(ans,0,a3*a3+4*a6);
  return ans;
}

// div_pol_odd(a1,a2,a3,a4,a6,n) returns the coefficients of the
// polynomial in x whose zeros are the (x-coordinates of the) non-zero
// points P on E=[a1,a2,a3,a4,a6] satisfying nP=0 (odd n)

// The poly itself is found recursively

ZPoly div_pol_odd(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                  const bigint& a6, int n)
{
  static const bigint four(4);
  ZPoly X; ZPolySetX(X);
  ZPoly f1 = X*(X*(X+a2)+a4)+a6;
  ZPoly f2 = a1*X+a3;
  ZPoly psi24=(four*f1+f2*f2); psi24*=psi24;
  ZPoly ans;
  switch(n) {
  case 0: 
    SetDegree(ans,0); SetCoeff(ans,0,0); return ans;
  case 1: case 2: 
    SetDegree(ans,0); SetCoeff(ans,0,1); return ans;
  case 3:
    SetDegree(ans,4);
    SetCoeff(ans,4,3);
    SetCoeff(ans,3,a1*a1+4*a2);
    SetCoeff(ans,2,3*a1*a3+6*a4);
    SetCoeff(ans,1,3*a3*a3+12*a6);
    SetCoeff(ans,0,a1*a1*a6-a1*a3*a4+a2*a3*a3+4*a2*a6-a4*a4);
    return ans;
  case 4:
    SetDegree(ans,6);
    SetCoeff(ans,6,2);
    SetCoeff(ans,5,a1*a1+4*a2);
    SetCoeff(ans,4,5*a1*a3+10*a4);
    SetCoeff(ans,3,10*a3*a3+40*a6);
    SetCoeff(ans,2,10*a1*a1*a6-10*a1*a3*a4+10*a2*a3*a3+40*a2*a6-10*a4*a4);
    SetCoeff(ans,1,a1*a1*a1*a1*a6-a1*a1*a1*a3*a4+a1*a1*a2*a3*a3+8*a1*a1*a2*a6-a1*a1*a4*a4-4*a1*a2*a3*a4-a1*a3*a3*a3-4*a1*a3*a6+4*a2*a2*a3*a3+16*a2*a2*a6-4*a2*a4*a4-2*a3*a3*a4-8*a4*a6);
    SetCoeff(ans,0,a1*a1*a1*a3*a6-a1*a1*a3*a3*a4+2*a1*a1*a4*a6+a1*a2*a3*a3*a3+4*a1*a2*a3*a6-3*a1*a3*a4*a4+2*a2*a3*a3*a4+8*a2*a4*a6-a3*a3*a3*a3-8*a3*a3*a6-2*a4*a4*a4-16*a6*a6);
    return ans;
  default: // general case, use recursion
    // If n is odd, n=2m+1:
    if(n%2==1)
      {
	int m=(n-1)/2;
	ZPoly t1=div_pol_odd(a1,a2,a3,a4,a6,m);
	t1=div_pol_odd(a1,a2,a3,a4,a6,m+2)*t1*t1*t1;
	ZPoly t2=div_pol_odd(a1,a2,a3,a4,a6,m+1);
	t2=div_pol_odd(a1,a2,a3,a4,a6,m-1)*t2*t2*t2;
	if(m%2==1) return t1-psi24*t2;
	return psi24*t1-t2;
      }
    else // n is even, n=2m:
      {
	int m=n/2;
	ZPoly t1=div_pol_odd(a1,a2,a3,a4,a6,m-1);
	t1=div_pol_odd(a1,a2,a3,a4,a6,m+2)*t1*t1;
	ZPoly t2=div_pol_odd(a1,a2,a3,a4,a6,m+1);
	t2=div_pol_odd(a1,a2,a3,a4,a6,m-2)*t2*t2;
	return div_pol_odd(a1,a2,a3,a4,a6,m)*(t1-t2);
      }
  }
}

ZPoly div_pol(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
              const bigint& a6,int n)
{
  return (n==2? div_pol_2(a1,a2,a3,a4,a6) :  div_pol_odd(a1,a2,a3,a4,a6,n));
}

ZPoly division_polynomial(Curvedata* E, int p)
{
  bigint a1,a2,a3,a4,a6;
  E->getai(a1,a2,a3,a4,a6);
  return (p==2? div_pol_2(a1,a2,a3,a4,a6) :  div_pol_odd(a1,a2,a3,a4,a6,p));
}

// Numerator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_num(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                   const bigint& a6, int n)
{
  ZPoly X; ZPolySetX(X);
  ZPoly P_2 = div_pol_2(a1,a2,a3,a4,a6);
  ZPoly P_n = div_pol_odd(a1,a2,a3,a4,a6,n);
  ZPoly P_nplus1 = div_pol_odd(a1,a2,a3,a4,a6,n+1);
  ZPoly P_nminus1 = div_pol_odd(a1,a2,a3,a4,a6,n-1);
  ZPoly A = X * P_n * P_n;
  ZPoly B = P_nminus1 * P_nplus1;
  return (n%2==0? A * P_2 - B : A - P_2 * B);
}

// Denominator of the multiplication-by-n map on the x-coordinate

ZPoly mul_by_n_den(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                   const bigint& a6, int n)
{
  ZPoly P_n = div_pol_odd(a1,a2,a3,a4,a6,n);
  return (n%2==1? P_n * P_n : P_n * P_n * div_pol_2(a1,a2,a3,a4,a6));
}

// Polynomial whose roots are x(Q) for Q satisfying n*Q=P, where x(P)=xP/zP

ZPoly division_points_X_pol(const bigint& a1,const bigint& a2,const bigint& a3,const bigint& a4,
                            const bigint& a6,
                            int n,
                            const bigint& xP, const bigint& zP)
{
  ZPoly numpoly = mul_by_n_num(a1, a2, a3, a4, a6, n);
  ZPoly denpoly = mul_by_n_den(a1, a2, a3, a4, a6, n);
  //cout << "numpoly = " << numpoly << endl;
  //cout << "denpoly = " << numpoly << endl;
  return zP * numpoly - xP * denpoly;
}
