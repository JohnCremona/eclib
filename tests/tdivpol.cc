// tdivpol.cc -- test for division poly functions in divpol.h/cc 
//////////////////////////////////////////////////////////////////////////
//
// Copyright 1990-2023 John Cremona
// 
// This file is part of the eclib package.
// 
// eclib is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
// 
// eclib is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
// 
// You should have received a copy of the GNU General Public License
// along with eclib; if not, write to the Free Software Foundation,
// Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
// 
//////////////////////////////////////////////////////////////////////////
//

#include <eclib/curve.h>

#include <eclib/polys.h>
#include <eclib/divpol.h>
#include <eclib/points.h>

int main()
{
  initprimes("PRIMES",0);

  Curve E(BIGINT(0),BIGINT(0),BIGINT(1),BIGINT(-7),BIGINT(6));

  Curvedata C(E, 0);
  cout << "Curve " << E << endl;

  int i;
  cout << "\nDivision polynomials:\n\n";
  for(i=2; i<12; i+=1)
    cout<<"Division Poly ("<<i<<") = \t" << division_polynomial(&C,i) << endl;

  cout << "\nMultiplication-by-n maps (x coordinate):\n\n";

  bigint a1,a2,a3,a4,a6;
  E.getai(a1,a2,a3,a4,a6);
  vector<int> nn = {2,3,5,7};
  for(vector<int>::iterator n = nn.begin(); n!=nn.end(); n++)
    {
      i = *n;
      cout << "n = " << i << ":\n";
      cout<<"Numerator:  \t" << mul_by_n_num(a1,a2,a3,a4,a6,i) << endl;
      cout<<"Denominator:\t" << mul_by_n_den(a1,a2,a3,a4,a6,i) << endl;
    }

#if(1)
  cout << "\nTesting division of points:\n";

  Point P0(C, BIGINT(0),BIGINT(2)) ;
  Point P1(C, BIGINT(1),BIGINT(0)) ;
  Point P2(C, BIGINT(2),BIGINT(0)) ;

  vector<Point> Plist = {P0, P1, P2, P0+P1, P0+P2, P1+P2, P0+P1+P2};
  vector<int> mlist = {2,3,5,7,11,13};
  for (vector<Point>::iterator Pi = Plist.begin(); Pi!=Plist.end(); Pi++)
    {
      Point P = *Pi;
      cout << "\nP = " << P << endl;
      for (vector<int>::iterator mi = mlist.begin(); mi!=mlist.end(); mi++)
        {
          int m = *mi;
          Point Q = m*P;
          vector<Point> newP = Q.division_points(m);
          cout << m << "*P = " << Q << ", divided by " << m << " gives back ";
          if (newP.size()==1 && newP[0]==P)
            {
              cout << "P, OK" << endl;
            }
          else
            {
              cout << newP << "???" << endl;
            }
        }
    }
#endif
}


//end of file tdivpol.cc





