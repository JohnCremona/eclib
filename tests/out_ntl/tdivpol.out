Curve [0,0,1,-7,6]

Division polynomials:

Division Poly (2) = 	[25 -28 0 4]
Division Poly (3) = 	[-49 75 -42 0 3]
Division Poly (4) = 	[61 350 -490 250 -70 0 2]
Division Poly (5) = 	[155774 -406875 380975 -219125 224875 -213150 93525 -10500 -5145 2375 -434 0 5]
Division Poly (6) = 	[-7450597 33433150 -55698846 38175900 2995139 -22858500 16502416 -7083900 3172008 -1401400 336784 0 -17836 4200 -504 0 3]
Division Poly (7) = 	[-18468518451 129897313175 -445857037878 932404235875 -1243400007052 1026876435550 -437452731142 -17749822300 107968849198 -15732678675 -41963963776 32734719500 -10597551734 1215763500 204773548 -134374100 71781619 -28348950 5133611 4900 -144746 24650 -2156 0 7]
Division Poly (8) = 	[-2571419770853 1215506524050 67371048936842 -320148770148050 780063910616298 -1249129857057600 1423544884020198 -1160552647602600 621047252595462 -133390086758000 -94508756302632 102557214742500 -42735421158370 6650264571300 470955582384 859919970900 -1163784258588 520062744600 -101339901340 -2100672000 5910589524 -1768634400 442778784 -97461000 11262114 409500 -266364 31300 -2044 0 4]
Division Poly (9) = 	[184160099143680751 -1745915691101562000 7553596765573730970 -20009912719373528175 37942026581120864697 -60375121456041861525 93410183300843035368 -141970066026214424625 190405429869508518294 -205725048911535979875 170148544422339746088 -101903285497398603750 37479654863464097844 -301195540341584325 -10122360365346701220 7022704801725188100 -2024994639983255253 -378932954398234425 638476088702667243 -278350695029482875 46649147045567535 8235437040258750 -4808056523794875 -19863562351500 413123504333085 -21629335715325 -61645404898854 25482225972000 -4079567512557 -51564996000 145726971768 -34704720000 6686917839 -1118177550 97716879 3647700 -1573194 142650 -7308 0 9]
Division Poly (10) = 	[413009374512445941380 -5577533221799587495650 34236508339578868067654 -126719683386541187863300 312994894498933292419431 -530641300710685283735000 585897029187423197307976 -280834885094634332916300 -367204662156668126267015 1082254486519067947850050 -1554831039561822593605650 1672867489720306188031100 -1511763332284285767508693 1186278164953699558211100 -786873604739812654099112 405920703060332401917500 -130861468159990009717737 -5601080695755881466900 37964990852501696833520 -25524121253263387893600 8897497566342411821410 -784837740659180354300 -925670621559347454588 522167496232044339600 -83452626309791455327 -45657452461518723000 34482142926904321488 -10360740343811041800 1058457111228061160 328746903184114800 -137359646628361520 17449112283988000 -1181088726955137 1434498958414150 -782923178241638 188460706507500 -17341433546593 -2258719474900 951528824400 -162366029100 21994486360 -2572647000 159807824 8131200 -2124444 149000 -6104 0 5]
Division Poly (11) = 	[-54041658613892771450681999 880907080312611332584912000 -7219594509378023629107644470 38958100184313046991444736425 -152640690458238768754008318817 454785460748010763365906132375 -1056137061172539849022036113531 1936054859688651591523350738175 -2810594691494082952472648616520 3208473937931657698860198494875 -2822039231599463382188559156202 1848423918012109590080578699875 -912617258569632579061077286483 560913385968977938523311946800 -817656871503167956435977465356 1249509868334450532408690045625 -1414165966153163021643340433938 1199336430865617781090316458650 -784676191760539643225675692205 396756309706661409373382542875 -144956275362751028758501833067 23355448732213310557101714875 17220044440091899138064455473 -20247521386823417986151910750 11805434554239601412412797599 -4040005394287265921365582350 234083284571165543297755125 642823838857235347992405825 -419159192063093806076371388 127589838582618892736949000 -6470559713355987866471481 -12226425861215930742609000 5613930837990861266425006 -822808169809586975048075 -285218373979309991270589 186775240363627878171825 -38115245033833256003913 -4184634313923610398775 4784966662201425769762 -1316603611701382184125 126856023732928120092 24162795683139180375 -9281076375640161037 1097831192157003150 -100276466219262732 63186211155991825 -25032991453137052 4693816276634100 -310935739698815 -50338154356875 15290737879263 -2102529532125 227429915559 -21041940150 1025882033 51025100 -10176859 580525 -19558 0 11]

Multiplication-by-n maps (x coordinate):

n = 2:
Numerator:  	[49 -50 14 0 1]
Denominator:	[25 -28 0 4]
n = 3:
Numerator:  	[-1525 -4641 14700 -10473 1050 1470 -600 84 0 1]
Denominator:	[2401 -7350 9741 -6300 1470 450 -252 0 9]
n = 5:
Numerator:  	[11362160425 25747089400 -441542258500 1506059616215 -2610427772875 2620690440081 -1387765627500 -5048462735 622380767750 -514545267425 213273944550 -37895439750 -6392442000 5509239565 -1991083500 805974932 -259821875 26485165 8844500 -2116075 -204050 110250 -13000 700 0 1]
Denominator:	[24265539076 -126761092500 284239264925 -378286361750 393514275875 -416360781200 421947611075 -340338334000 222184587415 -139925348500 86109315418 -40174487500 9539139565 1483576500 -2055966500 735111700 -102334925 -17456250 11041735 -2166500 136906 23750 -4340 0 25]
n = 7:
Numerator:  	[-478965310761451231025 3253200126164286188149 493504842253691880150 -87437670391256256814141 481053400926406685558475 -1474524847437726457036608 3051293047895397342722300 -4507790169717305498341981 4708354723086060377511600 -3031718411072855187719700 100320809977898287428700 2584762647056569539994518 -3803952756150941424717400 3428800794002557340072846 -2220821754300618338952600 1043523046720703474603804 -312300468078508586572175 1442623950439048333404 74652464685011903431000 -62420322870375191780779 32723404759644607191750 -10817148832748217403500 983621853556791774500 1094908222340954435400 -597643354403853229500 85498396077199034566 40785097726864444800 -20343885839157086724 223835554725509900 2681079531989243980 -1026461532749354800 119137846302349868 37877316216756800 -18977605618891227 3304306659173450 8701877287670 -135764897535400 34412538727721 -5354282777900 727182801268 -86238689250 4295207210 519821400 -43847846 -10392900 1754788 -98000 2744 0 1]
Denominator:	[341086173774927439401 -4798021850215625783850 33341949831200126354581 -150271512229128794144550 486949620236074001282038 -1192396394453621322632800 2261070134715425544591521 -3347375281558187150583900 3842454854008922612126960 -3324943281677884470444750 2010418069392193707136282 -651020404345562013303800 -96608621157166208197954 166151917069487225505300 48699725372809023988580 -178444916497903457931200 144560299534716328862298 -54814088180206102522950 961202166802238932207 8616789704135225396700 -2291677169592145859020 -1813626803872896374350 1838081104627613478024 -798819036672743394000 191378585118824669566 -10013444688503331600 -15500999700626997900 10243440676476617400 -4123480548928793500 1066016998482282000 -119062576319787996 -30000425318197600 19749888363614829 -6463294466493550 1585109901891446 -246150026516100 -2360054651479 12392379246100 -3190374052068 372027946500 61786550 -7553991900 1303637806 -106222200 2621892 345100 -30184 0 49]

Testing division of points:

P = [0:2:1]
2*P = [245:-32:125], divided by 2 gives back P, OK
3*P = [-74725:-438957:117649], divided by 3 gives back P, OK
5*P = [1769929178043950:4738372533007263:3779940084024824], divided by 5 gives back P, OK
7*P = [-8845779679186810919721876142275:-26132116500093837448172183040908:6299356293743239735763552887851], divided by 7 gives back P, OK
11*P = [-345907035865394112110113258379175841526681422618969349065465631483788171611500:-603899496397932670693901238084339680074376067505653656300679516648438378729902:157828710767928051843902784106074829349824572090398507020120360018086980045999], divided by 11 gives back P, OK
13*P = [4330944971498881477952763040195461194737358561970732874089910688931368534665435456142865934677314444316324525:474166894211156940132146040087902301358254640922900664396468316978550001533595729712356829915629891398826393:4459345156448775844130923624171028485455804112741895938417180204973106756635206294993276819983272953020126899], divided by 13 gives back P, OK

P = [1:0:1]
2*P = [14:51:1], divided by 2 gives back P, OK
3*P = [858:3056:2197], divided by 3 gives back P, OK
5*P = [-13422227300:37155830557:12167000000], divided by 5 gives back P, OK
7*P = [-654363158444498413938:323334584599779613133:232135196336821724283], divided by 7 gives back P, OK
11*P = [-5386909266611602426586636855741750187975628118240481:-31903410988726312174342705633718553251007282422752800:8563273080452283301577237101778658433493825993941319], divided by 11 gives back P, OK
13*P = [6790601027182872441717735378318194615000557734419531444518829576329287422:-21346190491686533430108554189028709484678265994449397859310470032693638296:10895308449429756233096859998848382387567744114034063749875541435853032123], divided by 13 gives back P, OK

P = [2:0:1]
2*P = [21:-96:1], divided by 2 gives back P, OK
3*P = [17347:11477:6859], divided by 3 gives back P, OK
5*P = [1314368321100:2192807253559:296474109632], divided by 5 gives back P, OK
7*P = [787735371750610845962493:2815013979731975184348416:58266402897702006018621], divided by 7 gives back P, OK
11*P = [465224049592215941888315944477150086362582688911509781121082:-2835041364493288591742476527691666698801289699650164214285000:12520837463201904227545769580855021289721316529527142889041], divided by 11 gives back P, OK
13*P = [967652288848675141165417507105084877050922452157366640667361749659468531276841232963:-2467719925245805762134421012865078764322728240276925184295644784012954762439103997267:137922062795787281492160592505828749775362042044145821626654801172360654848016552211], divided by 13 gives back P, OK

P = [3:3:1]
2*P = [742:-552:343], divided by 2 gives back P, OK
3*P = [1735612:-8697671:68921], divided by 3 gives back P, OK
5*P = [13274790938560307650:-3810346189413166609:6816655498982663368], divided by 5 gives back P, OK
7*P = [15610084384521543609511648390658094288:68045384727563741553482631637942786750:797803639727363681600463977477765079], divided by 7 gives back P, OK
11*P = [1575803501107696145915932812044229184495312851311623153876318357968752759104183973347056429397:1396274583140038254949640133020198974959304566276876829967194082976845151053935966087848230097:557829383989248411534035078306591098535924517768571182260197731835239742103251080379853060699], divided by 11 gives back P, OK
13*P = [6161146492714921360750727810183946205467564656490939099032408800798722540272737508766984883606134682101989119243397164919137164312:-35721620630687298226865205673450160556524720425884654664577788075243215491371022249911678723584335179331885213925899777407539593821:183117292199220258281649737393844871119425437622572982949683718143218212229910662967343636339932171621634631574337220088031542071], divided by 13 gives back P, OK

P = [-1:-4:1]
2*P = [798:377:343], divided by 2 gives back P, OK
3*P = [4430992:-438330:4330747], divided by 3 gives back P, OK
5*P = [-12221903491172643150:11468039763290751651:4885740073494125000], divided by 5 gives back P, OK
7*P = [17650003281779675566729443347938708188:-192526580513012204672730365490691444763:72980796628288414077404408180857060533], divided by 7 gives back P, OK
11*P = [-30613130525745975806945896359562605604200966296242976777041809153479858248968536822155766662519:-15714745521957622294462788054133735354109125384042241494977265337200555303010618130240791627276:10342901057480405091454441780213283779071882904636290875244915407676321355926935920245292819619], divided by 11 gives back P, OK
13*P = [1405907707042133954154238850941035663903445083801514062880744416997628088406667394648913262927105511332783388554595372344393472664228:-2211226265150483927592113283252013630287688089685314399591313107761425853236835065390795882450722392047791605464694483959070884011570:1600088061666351860094194147994818358340122989198013456481419612192269188713359699781782586135291445838482977423277673449499162776373], divided by 13 gives back P, OK

P = [-3:-1:1]
2*P = [406:8180:1], divided by 2 gives back P, OK
3*P = [-198562138:-133849341:68417929], divided by 3 gives back P, OK
5*P = [-868891893753069223266400:-892775925650647678206441:320035512450753499922432], divided by 5 gives back P, OK
7*P = [-250789195524699438713028256143149165935010703462:-351763092836663361098186249071565146814058578612:102208756706176009076316662108042042887248607671], divided by 7 gives back P, OK
11*P = [-5888498972561699214345998647776423563516419612163855732932875428098479096816453556740071283543529303920465099353451197:-13533256227761692733082484218341561251156460176136586962103570474399526108120718796397535148131526130064613929899679999:3291138473665024086163024275753883782780254080619125033356991189541952502966715638923169363573135424125729714408703999], divided by 11 gives back P, OK
13*P = [-471304787040968158056679651141156035898503824031997613220205307381493602197020538234554557851623914285000405516091920163285632114952834039661937557168333773226984038:-1370873330960135587145910098464877803541908169825891341967548989668552498702749973686005785221486327913645312167741302000549449734377969308776534759994492228417393091:330212063553055881533947376167502836269069675579486301004644611488773223934819786676041306852236930427317775633620010841259816717836527910178288832111727833861202679], divided by 13 gives back P, OK

P = [4:-7:1]
2*P = [4277:-1045:2197], divided by 2 gives back P, OK
3*P = [173402493:270630132:41781923], divided by 3 gives back P, OK
5*P = [347958005352980994267950:-591491125266464506985649:90144087121871684875000], divided by 5 gives back P, OK
7*P = [28346791926960812679644342662875487909944722853:46007717633420125671884939882166416087274190645:6574733970234508106612925608541549400717628583], divided by 7 gives back P, OK
11*P = [17474145883556647766828376232602106710826303890390986982624317936592550275465510611644786949719563721070184568269144:29480256355813074498457852565032175239305961624890346125612405835647274977712834264461974502384623321822021473979573:3896217090865641720447450673233504443391236048556578877575563503584301366068007409294978657913670393568854035781361], divided by 11 gives back P, OK
13*P = [141214430924449205970890907019814623022662416842585435331280731630294938794541398360467356587791564443343502605886099748817931388768459086946049432514093725256267:-226503185172742643796701194535524557733787205454975272078361208779360866680687110556127834139290657225066635029141642546627353273657327903648741011104842867737892:39145148418716729635148349998510974266838033447983885103993474847107471829372688807853030441396426545342852451854525365130399276623238448918996402000657968479137], divided by 13 gives back P, OK
